### PR TITLE
Track email signup user journey

### PR DESCRIPTION
**Pull request by laurentqro**
Adds analytics tracking to email sign-up journey.

https://trello.com/c/mTcA2hFA/307-add-tracking-to-email-sign-up-journey

https://docs.google.com/presentation/d/1sPjnEiA9w7Ctrl-7HLkOj9WNwJ_6ShnP3j7g06JS7n0/edit?ts=5efc8fc7#slide=id.p

Uses the GDS way module structure, as per https://gds-way.cloudapps.digital/manuals/programming-languages/js.html#modules.

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -0,0 +1,878 @@
+{
```
Comment (alex-ju) [inline package-lock.json:1]: Really appreciate setting up Standard for listing 🙇  We're trying to use yarn in Rails projects – as I understand it comes out of the box. There's a good reference for the [package.json in content-publisher](https://github.com/alphagov/content-publisher/blob/master/package.json).

File: Gemfile.lock [hunk 2]
```diff
@@ -121,6 +121,15 @@ GEM
       domain_name (~> 0.5)
     i18n (1.8.3)
       concurrent-ruby (~> 1.0)
+    jasmine (3.5.1)
+      jasmine-core (~> 3.5.0)
+      phantomjs
```
[the remark this answers is not in the document]
Comment (kevindew) [reply Gemfile.lock:126]: Jasmine bundles phantomjs unfortunately. We can't avoid it being a dependency 😞  we [end up with it in Content Publisher too](https://github.com/alphagov/content-publisher/blob/d41fabb5ce00f137ed6b409c130c51c499f91b9f/Gemfile.lock#L193). Might not be something resolved until we reach the distant land of webpacker and jest/karma.

File: app/assets/javascripts/application.js [hunk 3]
```diff
@@ -1,6 +1,9 @@
+//= require govuk_publishing_components/dependencies
```
[the remark this answers is not in the document]
Comment (kevindew) [reply app/assets/javascripts/application.js:1]: I recall this being needed for the unit tests on a comment on the previous PR.

The approach you can take to include this in only tests is to add it to the start of jasmine.yml src definition. For example:

```yml
src_files:
  - assets/govuk_publishing_components/dependencies.js
  - assets/application.js
```

and then would also need referencing in config/initializers/assets.rb to flag it as a file that needs precompiling.

File: Gemfile [hunk 4]
```diff
@@ -16,6 +16,8 @@ gem "uglifier", "~> 4.2"
 
 group :development, :test do
   gem "govuk_test", "~> 1.0"
+  gem "jasmine", "~> 3.5.1"
+  gem "jasmine_selenium_runner", "~> 3", require: false
```
Comment (kevindew) [inline Gemfile:20]: Sorry to be a pain again, we pushed up a new approach to pulling in jasmine_selenium_runner that avoids the need for `require: false` that we're going to try popularise before the `require: false` technique spreads too far (as it essentially fights the gems configuration). Details are here: https://github.com/alphagov/content-publisher/pull/2110

File: app/assets/javascripts/modules/track-email-alert-signup-choices.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    }
+  }
+
+  TrackEmailAlertSignupChoices.prototype.forwardToAnalytics = function () {
+    if (!GOVUK.analytics || !GOVUK.analytics.trackEvent) { return }
+    return GOVUK.analytics.trackEvent.apply(GOVUK, arguments)
+  }
+
+  Modules.TrackEmailAlertSignupChoices = TrackEmailAlertSignupChoices
+})(window.GOVUK.Modules)
```
Comment (kevindew) [inline app/assets/javascripts/modules/track-email-alert-signup-choices.js:46]: This looks great 👍

File: package.json [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  "directories": {
+    "doc": "docs",
+    "lib": "lib"
+  },
+  "dependencies": {},
+  "devDependencies": {
+    "eslint": "^7.3.1"
+  },
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1",
```
Comment (kevindew) [inline package.json:15]: Probably worth removing this line.

File: package.json [hunk 7]
```diff
@@ -0,0 +1,28 @@
+{
```
Comment (kevindew) [inline package.json:1]: We should mark this with private: true to indicate it's not a public package.

File: package.json [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+  "name": "email-alert-frontend",
+  "version": "1.0.0",
+  "description": "A frontend for creating and managing email subscriptions.",
```
Comment (kevindew) [inline package.json:4]: Somewhat moot given this is private but should reference GOV.UK in the description.

File: package.json [hunk 9]
```diff
@@ -0,0 +1,28 @@
+{
+  "name": "email-alert-frontend",
+  "version": "1.0.0",
```
Comment (kevindew) [inline package.json:3]: I don't think we want this since we won't maintain it (I believe it's still valid without a version)

File: package.json [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+  "description": "A frontend for creating and managing email subscriptions.",
+  "main": "index.js",
+  "directories": {
+    "doc": "docs",
+    "lib": "lib"
+  },
```
Comment (kevindew) [inline package.json:9]: I don't think we want/need the main and directories references here.

File: package.json [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+  "dependencies": {},
+  "devDependencies": {
+    "eslint": "^7.3.1"
```
Comment (kevindew) [inline package.json:12]: Should this actually be a dependency on [standard](https://www.npmjs.com/package/standard)?

File: package.json [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1",
+    "lint": "standard app/assets/javascripts/{*.js,**/*.js} spec/javascripts/{*.js,**/*.js}"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/alphagov/email-alert-frontend.git"
+  },
+  "author": "GDS",
+  "license": "ISC",
```
Comment (kevindew) [inline package.json:23]:
```suggestion
  "license": "MIT",
```

as per: https://github.com/alphagov/email-alert-frontend/blob/master/LICENCE

File: package.json [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+  },
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1",
+    "lint": "standard app/assets/javascripts/{*.js,**/*.js} spec/javascripts/{*.js,**/*.js}"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/alphagov/email-alert-frontend.git"
+  },
+  "author": "GDS",
```
Comment (kevindew) [inline package.json:22]:
```suggestion
  "author": "Government Digital Service",
```

File: package.json [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+  "license": "ISC",
+  "bugs": {
+    "url": "https://github.com/alphagov/email-alert-frontend/issues"
+  },
+  "homepage": "https://github.com/alphagov/email-alert-frontend#readme"
```
Comment (kevindew) [inline package.json:27]: I don't think we want/need these bugs and homepage links

File: Rakefile [hunk 15]
```diff
@@ -6,4 +6,4 @@ require File.expand_path("config/application", __dir__)
 Rails.application.load_tasks
 
 Rake::Task[:default].clear unless Rails.env.production?
-task default: %i[lint spec]
+task default: %i[lint spec jasmine:ci]
```
Comment (kevindew) [inline Rakefile:9]: We should also expand the lint task to include the JS linting, example: https://github.com/alphagov/content-publisher/blob/d41fabb5ce00f137ed6b409c130c51c499f91b9f/lib/tasks/lint.rake#L6

File: Gemfile.lock (hunk 2 above)
[the remark this answers is not in the document]
Comment (alex-ju) [reply Gemfile.lock:126]: Ah, right. Sorry about the confusion then.

File: app/assets/javascripts/application.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (laurentqro) [reply app/assets/javascripts/application.js:1]: This works, thanks @kevindew ! :tada:

**Comment by laurentqro** [comment]
> Great, looks good to me. Might be worth confirming with a PA that it works like they expect in integration before merging.

Tested on integration and approved by Andrew. 👌

**Pull request merged**